### PR TITLE
ddl: speed up add index with split tables

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/26683 

Problem Summary:
The SHARD_ROW_ID_BITS value is set or the table is SPLIT TABLE operated and a large amount of data is inserted into the table. Later, adding an index to the table, in a few cases, cause the backfill index to be slower than the write speed of the data in the table.

### What is changed and how it works?

Proposal: [xxx](url) 

How it Works:
In the original method, each worker originally adds indexes to the region range [startHandle, endHandle] by batch. In the new method, we get the range of actual data in each region to add indexes on a batch basis.

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

Start cluster with `tiup playground --tiflash 0` in local.

### case 1
```
create table tbl(a bigint, b bigint, b bigint) SHARD_ROW_ID_BITS = 4 pre_split_regions = 4;
```
Add 2 million rows to this table.
Then connect1 inserts 20 million rows(50 goroutines * 400 k), connect2 executes alter table tbl add index idx(a);.

Before this PR(**connect1 finished insert operation, 14 min 33 sec**):
```
tidb> alter table tbl add index idx(c1);
Query OK, 0 rows affected (22 min 58.35 sec)
```

After this PR(**add index backfill 6 million rows**):
```
tidb> alter table tbl add index idx(c1);
Query OK, 0 rows affected (12 min 52.16 sec)
```
### case 2
```
CREATE TABLE tbl(a bigint PRIMARY KEY AUTO_RANDOM(4), b varchar(255), c bigint);
SPLIT TABLE tbl BETWEEN (-9223372036854775808) AND (9223372036854775807) REGIONS 16;
```
Add 2 million rows to this table.
Then connect1 inserts 5 million rows(50 goroutines * 100 k), connect2 executes alter table tbl add index idx(a);.

Before this PR(**connect1 finished insert operation, 4 min 20 sec**):
```
tidb> alter table tbl add index idx(c1);
Query OK, 0 rows affected (4 min 54.72 sec)
```

After this PR(**connect1 finished insert operation, 3 min 21 sec**):
```
tidb> alter table tbl add index idx(c1);
Query OK, 0 rows affected (6 min 40.69 sec)
```
In theory, this PR doesn't speed up the case particularly well, not as well as the previous case.

### Release note 

```release-note
Optimize ”Add Index” for an explicit split region table 
```

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @crazycs520 @tangenta @AilinKid

File: ddl/db_test.go [hunk 1]
```diff
@@ -1401,6 +1452,105 @@ LOOP:
...
+			// So here is a limit to the number of rows inserted.
+			if num >= 1000 {
+				break
+			}
+			step := 50
+			// delete, insert and update some data
+			for i := num; i < num+step; i++ {
+				sql := fmt.Sprintf("delete from test_add_index where a = %d", i+1)
+				tk.MustExec(sql)
+				sql = fmt.Sprintf("insert into test_add_index values (%d, %d, %d)", i, i, i)
```
Comment (crazycs520) [inline ddl/db_test.go:1542]: may get bellow error:

```log
    tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
... value *errors.withStack = [ddl:8216]Invalid auto random: Explicit insertion on auto_random column is disabled. Try to set @@allow_auto_random_explicit_insert = true. ("[ddl:8216]Invalid auto random: Explicit insertion on auto_random column is disabled. Try to set @@allow_auto_random_explicit_insert = true.")
... sql:insert into test_add_index values (1, 1, 1), [], error stack [ddl:8216]Invalid auto random: Explicit insertion on auto_random column is disabled. Try to set @@allow_auto_random_explicit_insert = true.

```

File: ddl/backfilling.go [hunk 2]
```diff
@@ -444,17 +444,27 @@ func tryDecodeToHandleString(key kv.Key) string {
...
 	physicalTableID := reorgInfo.PhysicalTableID
 
 	// Build reorg tasks.
 	for _, keyRange := range kvRanges {
+		endKey := keyRange.EndKey
+		endK, err := getRangeEndKey(workers[0].sessCtx.GetStore(), workers[0].priority, t, keyRange.StartKey, endKey)
+		if err != nil {
+			logutil.BgLogger().Info("[ddl] send range task to workers, get reverse key failed", zap.Error(err))
+		} else {
+			logutil.BgLogger().Info("[ddl] send range task to workers, change end key",
```
Comment (crazycs520) [inline ddl/backfilling.go:459]: Current log is a little bit hard to read:

```
[2021/08/05 19:31:11.821 +08:00] [INFO] [backfilling.go:459] ["[ddl] send range task to workers, change end key"] ["end key"="t\ufffd\u0000\u0000\u0000\u0000\u0000\u00008_r\ufffd\u0000\u0000\u0000\u0000\u0000\u0000\u0000"] ["current end key"="t\ufffd\u0000\u0000\u0000\u0000\u0000\u00008_r\ufffd\u0000\u0000\u0000\u0000\u0005\ufffd\ufffd"]
```

How about use `tryDecodeToHandleString`, the log will be:

```log
[2021/08/05 19:34:49.797 +08:00] [INFO] [backfilling.go:459] ["[ddl] send range task to workers, change end key"] ["end key"=5188146770730811392] ["current end key"=4611686018427754381]

```

File: ddl/backfilling.go [hunk 3]
```diff
@@ -444,17 +444,27 @@ func tryDecodeToHandleString(key kv.Key) string {
...
 	totalAddedCount *int64, kvRanges []kv.KeyRange) ([]kv.KeyRange, error) {
 	batchTasks := make([]*reorgBackfillTask, 0, len(workers))
 	physicalTableID := reorgInfo.PhysicalTableID
 
 	// Build reorg tasks.
 	for _, keyRange := range kvRanges {
+		endKey := keyRange.EndKey
+		endK, err := getRangeEndKey(workers[0].sessCtx.GetStore(), workers[0].priority, t, keyRange.StartKey, endKey)
+		if err != nil {
+			logutil.BgLogger().Info("[ddl] send range task to workers, get reverse key failed", zap.Error(err))
```
Comment (tangenta) [inline ddl/backfilling.go:457]:
```suggestion
			logutil.BgLogger().Error("[ddl] send range task to workers, get reverse key failed", zap.Error(err))
```

**Review by crazycs520 — APPROVED** [review]
LGTM

File: ddl/backfilling.go [hunk 4]
```diff
@@ -444,17 +444,27 @@ func tryDecodeToHandleString(key kv.Key) string {
...
+		} else {
+			logutil.BgLogger().Info("[ddl] send range task to workers, change end key",
+				zap.String("end key", tryDecodeToHandleString(endKey)), zap.String("current end key", tryDecodeToHandleString(endK)))
+			endKey = endK
```
Comment (tangenta) [inline ddl/backfilling.go:461]: This log seems to be annoying if there are a lot of regions.

File: ddl/backfilling.go (hunk 3 above)
Comment (zimulala) [reply ddl/backfilling.go:457]: I think the error level is too high, or change it to "warn"？

File: ddl/backfilling.go (hunk 4 above)
Comment (crazycs520) [reply ddl/backfilling.go:461]: This is a new feature, and this log will be helpful for diagnosis in the future.

**Review by tangenta — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Comment by zimulala** [comment]
Add case2.

**Comment by glkappe** [comment]
need a label like ‘cheery-pick-4.0’

**Pull request merged**